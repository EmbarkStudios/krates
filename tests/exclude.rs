mod util;

use krates::PkgSpec;
use util::build;

#[test]
fn excludes_workspace_member() {
    let mut kb = krates::Builder::new();
    kb.exclude(std::iter::once("a".parse::<PkgSpec>().unwrap()));

    let grafs = build("all-features.json", kb).unwrap();

    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn excludes_dependencies() {
    let mut kb = krates::Builder::new();

    let pkg_ids = [
        "bitflags",
        "bumpalo:3.11.0",
        "https://github.com/rust-lang/crates.io-index#byteorder",
        "https://github.com/rust-lang/crates.io-index#ring:0.16.20",
        "https://github.com/alexcrichton/cc-rs#cc",
    ];

    kb.exclude(pkg_ids.iter().map(|id| id.parse::<PkgSpec>().unwrap()));

    let grafs = build("all-features.json", kb).unwrap();
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn no_roots() {
    let mut kb = krates::Builder::new();

    // To ease testing, we just remove leaves that have no dependencies
    // themselves
    let pkg_ids = ["a", "b", "c"];

    kb.exclude(pkg_ids.iter().map(|id| id.parse::<PkgSpec>().unwrap()));

    let contents = std::fs::read_to_string("tests/all-features.json")
        .map_err(|e| format!("failed to load metadata file: {}", e))
        .unwrap();

    let md: krates::cm::Metadata = serde_json::from_str(&contents)
        .map_err(|e| format!("failed to deserialize metadata: {}", e))
        .unwrap();

    match kb.build_with_metadata::<util::JustId, krates::Edge, _>(md, |_f: krates::cm::Package| {
        panic!("shouldn't get here")
    }) {
        Err(krates::Error::NoRootKrates) => {}
        _ => panic!("expected no root crates error!"),
    }
}
