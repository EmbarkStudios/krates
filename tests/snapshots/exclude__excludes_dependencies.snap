---
source: tests/exclude.rs
expression: grafs.dotgraph()
---
digraph {
    0 [ label = "crate a 0.1.0 (path+file:///home/jake/code/krates/tests/ws/a)" ]
    1 [ label = "crate b 0.1.0 (path+file:///home/jake/code/krates/tests/ws/b)" ]
    2 [ label = "crate bindgen 0.59.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    3 [ label = "crate c 0.1.0 (path+file:///home/jake/code/krates/tests/ws/c)" ]
    4 [ label = "crate cc 1.0.73 (registry+https://github.com/rust-lang/crates.io-index)" ]
    5 [ label = "crate cexpr 0.6.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    6 [ label = "crate cfg-if 0.1.10 (registry+https://github.com/rust-lang/crates.io-index)" ]
    7 [ label = "crate cfg-if 1.0.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    8 [ label = "crate clang-sys 1.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    9 [ label = "crate coreaudio-rs 0.9.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    10 [ label = "crate coreaudio-sys 0.2.10 (registry+https://github.com/rust-lang/crates.io-index)" ]
    11 [ label = "crate difference 2.0.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    12 [ label = "crate glob 0.3.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    13 [ label = "crate js-sys 0.3.60 (registry+https://github.com/rust-lang/crates.io-index)" ]
    14 [ label = "crate lazy_static 1.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    15 [ label = "crate lazycell 1.3.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    16 [ label = "crate leftpad 0.2.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    17 [ label = "crate libc 0.2.133 (registry+https://github.com/rust-lang/crates.io-index)" ]
    18 [ label = "crate libloading 0.7.3 (registry+https://github.com/rust-lang/crates.io-index)" ]
    19 [ label = "crate log 0.4.17 (registry+https://github.com/rust-lang/crates.io-index)" ]
    20 [ label = "crate memchr 2.5.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    21 [ label = "crate minimal-lexical 0.2.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    22 [ label = "crate nix 0.16.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    23 [ label = "crate nom 7.1.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    24 [ label = "crate once_cell 1.15.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    25 [ label = "crate peeking_take_while 0.1.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    26 [ label = "crate proc-macro2 1.0.44 (registry+https://github.com/rust-lang/crates.io-index)" ]
    27 [ label = "crate quote 1.0.21 (registry+https://github.com/rust-lang/crates.io-index)" ]
    28 [ label = "crate regex 1.6.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    29 [ label = "crate regex-syntax 0.6.27 (registry+https://github.com/rust-lang/crates.io-index)" ]
    30 [ label = "crate rustc-hash 1.1.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    31 [ label = "crate shlex 1.1.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    32 [ label = "crate spin 0.5.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    33 [ label = "crate syn 1.0.101 (registry+https://github.com/rust-lang/crates.io-index)" ]
    34 [ label = "crate unicode-ident 1.0.4 (registry+https://github.com/rust-lang/crates.io-index)" ]
    35 [ label = "crate void 1.0.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    36 [ label = "crate wasm-bindgen 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    37 [ label = "crate wasm-bindgen-backend 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    38 [ label = "crate wasm-bindgen-futures 0.4.33 (registry+https://github.com/rust-lang/crates.io-index)" ]
    39 [ label = "crate wasm-bindgen-macro 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    40 [ label = "crate wasm-bindgen-macro-support 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    41 [ label = "crate wasm-bindgen-shared 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    42 [ label = "crate web-sys 0.3.60 (registry+https://github.com/rust-lang/crates.io-index)" ]
    43 [ label = "crate winapi 0.2.8 (registry+https://github.com/rust-lang/crates.io-index)" ]
    44 [ label = "crate winapi 0.3.9 (registry+https://github.com/rust-lang/crates.io-index)" ]
    45 [ label = "crate winapi-i686-pc-windows-gnu 0.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    46 [ label = "crate winapi-x86_64-pc-windows-gnu 0.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    47 [ label = "feature default" ]
    48 [ label = "feature clang_6_0" ]
    49 [ label = "feature runtime" ]
    50 [ label = "feature std" ]
    51 [ label = "feature unicode" ]
    52 [ label = "feature default" ]
    53 [ label = "feature default" ]
    54 [ label = "feature default" ]
    55 [ label = "feature default" ]
    56 [ label = "feature Crypto" ]
    57 [ label = "feature Window" ]
    58 [ label = "feature std" ]
    59 [ label = "feature audio_toolbox" ]
    60 [ label = "feature audio_unit" ]
    61 [ label = "feature core_audio" ]
    62 [ label = "feature core_midi" ]
    63 [ label = "feature open_al" ]
    64 [ label = "feature runtime" ]
    65 [ label = "feature default" ]
    66 [ label = "feature errhandlingapi" ]
    67 [ label = "feature libloaderapi" ]
    68 [ label = "feature extra_traits" ]
    69 [ label = "feature default" ]
    70 [ label = "feature default" ]
    71 [ label = "feature std" ]
    72 [ label = "feature std" ]
    73 [ label = "feature proc-macro" ]
    74 [ label = "feature unicode" ]
    75 [ label = "feature unicode-age" ]
    76 [ label = "feature unicode-bool" ]
    77 [ label = "feature unicode-case" ]
    78 [ label = "feature unicode-gencat" ]
    79 [ label = "feature unicode-perl" ]
    80 [ label = "feature unicode-script" ]
    81 [ label = "feature unicode-segment" ]
    82 [ label = "feature proc-macro" ]
    83 [ label = "feature spans" ]
    84 [ label = "feature default" ]
    85 [ label = "feature default" ]
    86 [ label = "feature default" ]
    87 [ label = "feature full" ]
    88 [ label = "feature default" ]
    89 [ label = "feature MessageEvent" ]
    90 [ label = "feature Worker" ]
    91 [ label = "feature spans" ]
    92 [ label = "feature visit" ]
    93 [ label = "feature spans" ]
    94 [ label = "feature leftpad" ]
    95 [ label = "feature leftier-strings" ]
    96 [ label = "feature lazy_static" ]
    97 [ label = "feature libloading" ]
    98 [ label = "feature clang_5_0" ]
    99 [ label = "feature clang_4_0" ]
    100 [ label = "feature clang_3_9" ]
    101 [ label = "feature clang_3_8" ]
    102 [ label = "feature clang_3_7" ]
    103 [ label = "feature clang_3_6" ]
    104 [ label = "feature clang_3_5" ]
    105 [ label = "feature open_al" ]
    106 [ label = "feature audio_toolbox" ]
    107 [ label = "feature audio_unit" ]
    108 [ label = "feature core_audio" ]
    109 [ label = "feature core_midi" ]
    110 [ label = "feature std" ]
    111 [ label = "feature alloc" ]
    112 [ label = "feature std" ]
    113 [ label = "feature alloc" ]
    114 [ label = "feature race" ]
    115 [ label = "feature unicode-segment" ]
    116 [ label = "feature unicode-script" ]
    117 [ label = "feature unicode-perl" ]
    118 [ label = "feature unicode-gencat" ]
    119 [ label = "feature unicode-case" ]
    120 [ label = "feature unicode-bool" ]
    121 [ label = "feature unicode-age" ]
    122 [ label = "feature std" ]
    123 [ label = "feature std" ]
    124 [ label = "feature quote" ]
    125 [ label = "feature proc-macro" ]
    126 [ label = "feature printing" ]
    127 [ label = "feature parsing" ]
    128 [ label = "feature derive" ]
    129 [ label = "feature clone-impls" ]
    130 [ label = "feature std" ]
    131 [ label = "feature std" ]
    132 [ label = "feature spans" ]
    133 [ label = "feature EventTarget" ]
    134 [ label = "feature Event" ]
    0 -> 1 [ label = "" ]
    0 -> 47 [ label = "(dev)" ]
    0 -> 47 [ label = "(build) 'cfg(target_os = \"linux\")'" ]
    1 -> 47 [ label = "" ]
    1 -> 38 [ label = " 'cfg(all(target_arch = \"wasm32\", target_feature = \"atomics\"))'" ]
    1 -> 38 [ label = " 'cfg(all(target_vendor = \"xboxone\"))'" ]
    2 -> 5 [ label = "" ]
    2 -> 48 [ label = "" ]
    2 -> 49 [ label = "" ]
    2 -> 14 [ label = "" ]
    2 -> 15 [ label = "" ]
    2 -> 25 [ label = "" ]
    2 -> 26 [ label = "" ]
    2 -> 27 [ label = "" ]
    2 -> 50 [ label = "" ]
    2 -> 51 [ label = "" ]
    2 -> 52 [ label = "" ]
    2 -> 53 [ label = "" ]
    3 -> 54 [ label = " 'x86_64-apple-darwin'" ]
    3 -> 55 [ label = "(dev)" ]
    3 -> 14 [ label = " 'cfg(any(target_os = \"android\", target_os = \"freebsd\", target_os = \"linux\", target_os = \"netbsd\", target_os = \"openbsd\", target_os = \"solaris\"))'" ]
    3 -> 16 [ label = "" ]
    3 -> 17 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    3 -> 22 [ label = " 'x86_64-unknown-linux-gnu'" ]
    3 -> 32 [ label = " 'cfg(all(any(target_arch = \"aarch64\", target_arch = \"arm\", target_arch = \"x86\", target_arch = \"x86_64\"), not(target_os = \"ios\")))'" ]
    3 -> 56 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    3 -> 57 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    3 -> 43 [ label = " 'cfg(target_os = \"windows\")'" ]
    5 -> 58 [ label = "" ]
    8 -> 12 [ label = "" ]
    8 -> 12 [ label = "(build)" ]
    8 -> 17 [ label = "" ]
    8 -> 18 [ label = "" ]
    9 -> 10 [ label = "" ]
    9 -> 59 [ label = "" ]
    9 -> 60 [ label = "" ]
    9 -> 61 [ label = "" ]
    9 -> 62 [ label = "" ]
    9 -> 63 [ label = "" ]
    10 -> 64 [ label = "(build)" ]
    13 -> 65 [ label = "" ]
    18 -> 7 [ label = " 'cfg(unix)'" ]
    18 -> 66 [ label = " 'cfg(windows)'" ]
    18 -> 67 [ label = " 'cfg(windows)'" ]
    19 -> 7 [ label = "" ]
    22 -> 4 [ label = "(build) 'cfg(target_os = \"dragonfly\")'" ]
    22 -> 6 [ label = "" ]
    22 -> 68 [ label = "" ]
    22 -> 69 [ label = "" ]
    22 -> 70 [ label = "" ]
    23 -> 20 [ label = "" ]
    23 -> 71 [ label = "" ]
    23 -> 21 [ label = "" ]
    23 -> 72 [ label = "" ]
    26 -> 34 [ label = "" ]
    27 -> 26 [ label = "" ]
    27 -> 73 [ label = "" ]
    28 -> 29 [ label = "" ]
    28 -> 74 [ label = "" ]
    28 -> 75 [ label = "" ]
    28 -> 76 [ label = "" ]
    28 -> 77 [ label = "" ]
    28 -> 78 [ label = "" ]
    28 -> 79 [ label = "" ]
    28 -> 80 [ label = "" ]
    28 -> 81 [ label = "" ]
    33 -> 26 [ label = "" ]
    33 -> 73 [ label = "" ]
    33 -> 27 [ label = "" ]
    33 -> 82 [ label = "" ]
    33 -> 34 [ label = "" ]
    36 -> 7 [ label = "" ]
    36 -> 39 [ label = "" ]
    36 -> 83 [ label = "" ]
    37 -> 19 [ label = "" ]
    37 -> 84 [ label = "" ]
    37 -> 85 [ label = "" ]
    37 -> 86 [ label = "" ]
    37 -> 87 [ label = "" ]
    37 -> 88 [ label = "" ]
    37 -> 41 [ label = "" ]
    38 -> 7 [ label = "" ]
    38 -> 13 [ label = "" ]
    38 -> 65 [ label = "" ]
    38 -> 89 [ label = " 'cfg(target_feature = \"atomics\")'" ]
    38 -> 90 [ label = " 'cfg(target_feature = \"atomics\")'" ]
    39 -> 86 [ label = "" ]
    39 -> 40 [ label = "" ]
    39 -> 91 [ label = "" ]
    40 -> 85 [ label = "" ]
    40 -> 86 [ label = "" ]
    40 -> 92 [ label = "" ]
    40 -> 87 [ label = "" ]
    40 -> 88 [ label = "" ]
    40 -> 37 [ label = "" ]
    40 -> 93 [ label = "" ]
    40 -> 41 [ label = "" ]
    42 -> 13 [ label = "" ]
    42 -> 65 [ label = "" ]
    44 -> 45 [ label = " 'i686-pc-windows-gnu'" ]
    44 -> 46 [ label = " 'x86_64-pc-windows-gnu'" ]
    64 -> 2 [ label = "" ]
    64 -> 49 [ label = "" ]
    94 -> 3 [ label = "" ]
    94 -> 16 [ label = "" ]
    95 -> 3 [ label = "" ]
    95 -> 94 [ label = "" ]
    96 -> 3 [ label = "" ]
    96 -> 14 [ label = "" ]
    47 -> 3 [ label = "" ]
    47 -> 94 [ label = "" ]
    49 -> 8 [ label = "" ]
    49 -> 97 [ label = "" ]
    97 -> 8 [ label = "" ]
    97 -> 18 [ label = "" ]
    48 -> 8 [ label = "" ]
    48 -> 98 [ label = "" ]
    98 -> 8 [ label = "" ]
    98 -> 99 [ label = "" ]
    99 -> 8 [ label = "" ]
    99 -> 100 [ label = "" ]
    100 -> 8 [ label = "" ]
    100 -> 101 [ label = "" ]
    101 -> 8 [ label = "" ]
    101 -> 102 [ label = "" ]
    102 -> 8 [ label = "" ]
    102 -> 103 [ label = "" ]
    103 -> 8 [ label = "" ]
    103 -> 104 [ label = "" ]
    104 -> 8 [ label = "" ]
    105 -> 9 [ label = "" ]
    105 -> 63 [ label = "" ]
    54 -> 9 [ label = "" ]
    54 -> 106 [ label = "" ]
    54 -> 107 [ label = "" ]
    54 -> 108 [ label = "" ]
    54 -> 105 [ label = "" ]
    54 -> 109 [ label = "" ]
    109 -> 9 [ label = "" ]
    109 -> 62 [ label = "" ]
    108 -> 9 [ label = "" ]
    108 -> 61 [ label = "" ]
    107 -> 9 [ label = "" ]
    107 -> 60 [ label = "" ]
    106 -> 9 [ label = "" ]
    106 -> 59 [ label = "" ]
    63 -> 10 [ label = "" ]
    62 -> 10 [ label = "" ]
    61 -> 10 [ label = "" ]
    60 -> 10 [ label = "" ]
    59 -> 10 [ label = "" ]
    55 -> 11 [ label = "" ]
    110 -> 17 [ label = "" ]
    68 -> 17 [ label = "" ]
    69 -> 17 [ label = "" ]
    69 -> 110 [ label = "" ]
    71 -> 20 [ label = "" ]
    72 -> 21 [ label = "" ]
    58 -> 23 [ label = "" ]
    58 -> 111 [ label = "" ]
    58 -> 71 [ label = "" ]
    58 -> 72 [ label = "" ]
    111 -> 23 [ label = "" ]
    112 -> 24 [ label = "" ]
    112 -> 113 [ label = "" ]
    114 -> 24 [ label = "" ]
    84 -> 24 [ label = "" ]
    84 -> 112 [ label = "" ]
    113 -> 24 [ label = "" ]
    113 -> 114 [ label = "" ]
    73 -> 26 [ label = "" ]
    85 -> 26 [ label = "" ]
    85 -> 73 [ label = "" ]
    82 -> 27 [ label = "" ]
    82 -> 73 [ label = "" ]
    86 -> 27 [ label = "" ]
    86 -> 82 [ label = "" ]
    115 -> 28 [ label = "" ]
    115 -> 81 [ label = "" ]
    116 -> 28 [ label = "" ]
    116 -> 80 [ label = "" ]
    117 -> 28 [ label = "" ]
    117 -> 79 [ label = "" ]
    118 -> 28 [ label = "" ]
    118 -> 78 [ label = "" ]
    119 -> 28 [ label = "" ]
    119 -> 77 [ label = "" ]
    120 -> 28 [ label = "" ]
    120 -> 76 [ label = "" ]
    121 -> 28 [ label = "" ]
    121 -> 75 [ label = "" ]
    51 -> 28 [ label = "" ]
    51 -> 121 [ label = "" ]
    51 -> 120 [ label = "" ]
    51 -> 119 [ label = "" ]
    51 -> 118 [ label = "" ]
    51 -> 117 [ label = "" ]
    51 -> 116 [ label = "" ]
    51 -> 115 [ label = "" ]
    51 -> 74 [ label = "" ]
    50 -> 28 [ label = "" ]
    81 -> 29 [ label = "" ]
    80 -> 29 [ label = "" ]
    79 -> 29 [ label = "" ]
    78 -> 29 [ label = "" ]
    77 -> 29 [ label = "" ]
    76 -> 29 [ label = "" ]
    75 -> 29 [ label = "" ]
    74 -> 29 [ label = "" ]
    74 -> 75 [ label = "" ]
    74 -> 76 [ label = "" ]
    74 -> 77 [ label = "" ]
    74 -> 78 [ label = "" ]
    74 -> 79 [ label = "" ]
    74 -> 80 [ label = "" ]
    74 -> 81 [ label = "" ]
    122 -> 30 [ label = "" ]
    52 -> 30 [ label = "" ]
    52 -> 122 [ label = "" ]
    123 -> 31 [ label = "" ]
    53 -> 31 [ label = "" ]
    53 -> 123 [ label = "" ]
    92 -> 33 [ label = "" ]
    124 -> 33 [ label = "" ]
    124 -> 27 [ label = "" ]
    125 -> 33 [ label = "" ]
    125 -> 73 [ label = "" ]
    125 -> 82 [ label = "" ]
    126 -> 33 [ label = "" ]
    126 -> 124 [ label = "" ]
    127 -> 33 [ label = "" ]
    87 -> 33 [ label = "" ]
    128 -> 33 [ label = "" ]
    88 -> 33 [ label = "" ]
    88 -> 128 [ label = "" ]
    88 -> 127 [ label = "" ]
    88 -> 126 [ label = "" ]
    88 -> 129 [ label = "" ]
    88 -> 125 [ label = "" ]
    129 -> 33 [ label = "" ]
    130 -> 35 [ label = "" ]
    70 -> 35 [ label = "" ]
    70 -> 130 [ label = "" ]
    131 -> 36 [ label = "" ]
    132 -> 36 [ label = "" ]
    132 -> 83 [ label = "" ]
    65 -> 36 [ label = "" ]
    65 -> 132 [ label = "" ]
    65 -> 131 [ label = "" ]
    93 -> 37 [ label = "" ]
    83 -> 39 [ label = "" ]
    83 -> 91 [ label = "" ]
    91 -> 40 [ label = "" ]
    91 -> 93 [ label = "" ]
    90 -> 42 [ label = "" ]
    90 -> 133 [ label = "" ]
    57 -> 42 [ label = "" ]
    57 -> 133 [ label = "" ]
    89 -> 42 [ label = "" ]
    89 -> 134 [ label = "" ]
    133 -> 42 [ label = "" ]
    134 -> 42 [ label = "" ]
    56 -> 42 [ label = "" ]
    67 -> 44 [ label = "" ]
    66 -> 44 [ label = "" ]
}

